CenterlineMetrics/CenterlineMetrics.py

```diff
@@ -382,11 +443,91 @@ def getExtremeDiameterPoint(self, boolMaximum):
+  def getUnitNodeDisplayString(self, value, category):
     selectionNode = slicer.mrmlScene.GetNodeByID("vtkMRMLSelectionNodeSingleton")
-    unitNode = slicer.mrmlScene.GetNodeByID(selectionNode.GetUnitNodeID("length"))
+    unitNode = slicer.mrmlScene.GetNodeByID(selectionNode.GetUnitNodeID(category))
     return unitNode.GetDisplayStringFromValue(value)
     
+  
+  def createCrossSection(self, center, normal):
+    if self.segmentationNode is None or self.currentSegmentID == "":
+        # Don't print any message, is perhaps intended.
+        return
+    
+    util = Utility()
+    util.islandModelNode = self.islandModelNode
+    
+    # Place a plane perpendicular to the centerline
+    plane = vtk.vtkPlane()
+    plane.SetOrigin(center)
+    plane.SetNormal(normal)
+    
+    # Work on the segment's closed surface
+    closedSurfacePolyData = vtk.vtkPolyData()
+    self.segmentationNode.GetClosedSurfaceRepresentation(self.currentSegmentID, closedSurfacePolyData)
+    
+    # Though not needed here, it was surprising that vtkSegment::GetId() is missing.
+    currentSegment =  self.segmentationNode.GetSegmentation().GetSegment(self.currentSegmentID)
+    
+    # Cut through the closed surface and get the points of the contour.
+    planeCut = vtk.vtkCutter()
+    planeCut.SetInputData(closedSurfacePolyData)
+    planeCut.SetCutFunction(plane)
+    planeCut.Update()
+    planePoints = vtk.vtkPoints()
+    planePoints = planeCut.GetOutput().GetPoints()
+    # self.segmentationNode.GetDisplayNode().GetSegmentVisibility3D(self.currentSegmentID) doesn't work as expected. Even if the segment is hidden in 3D view, it returns True.
+    if planePoints is None:
+        msg = "Could not cut segment. Is it visible in 3D view?"
+        slicer.util.showStatusMessage(msg, 3000)
+        raise ValueError(msg)
+    if planePoints.GetNumberOfPoints() < 3:
+        logging.info("Not enough points to create surface")
+        return
+    
+    # Sort the points by proximity to retain those around the centre of the centerline
+    vCenter = [center[0], center[1], center[2]]
+    planePointsCopy = vtk.vtkPoints()
+    planePointsCopy.DeepCopy(planePoints)
+    # Empties planePointsCopy
+    sortedPoints = util.sortPointsByProximity(planePointsCopy, vCenter)
+    islandPoints = util.makeIslandAroundCenter(sortedPoints, center)
+    
+    # Make a model out of the points retained
+    islandPolyData = vtk.vtkPolyData()
+    # Prefer an inverted color for the model
+    segmentColor = currentSegment.GetColor()
+    crossSectionColor = [1 - segmentColor[0], 1 - segmentColor[1], 1 - segmentColor[2]]
+    # Finally create the model
+    self.currentSurfaceArea = self.xFitSurfaceProjectWarp(islandPoints, islandPolyData, crossSectionColor, currentSegment.GetName())
+
+  # Create an exact-fit model representing the cross-section.
+  def xFitSurfaceProjectWarp(self, islandPoints, islandPolyData, color = [1, 1, 0], segmentName = ""):
```

Comment:
We only need these complicated warpings if the curve is non-planar. If the curve is planar then https://vtk.org/doc/nightly/html/classvtkContourTriangulator.html is a much better choice. The main advantage of this filter is that it can handle not just concave regions but also if there are holes in the cross-section (e.g., you may have a small bubble in the segmentation). For proper handling of holes, you need to first triangulate the cross-section and then extract the part that is closest to the the centerline.